enarx-keep-sev/src/vm/load.rs

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use super::mem::AddressSpace;
+
+use std::io;
+
+use goblin::elf::program_header::{ProgramHeader, PT_LOAD};
+use x86_64::{PhysAddr, VirtAddr};
+
+fn parse_elf64(image: &[u8]) -> Result<goblin::elf::Elf, io::Error> {
+    let maybe_elf = goblin::Object::parse(&image)
+        .map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, e))?;
+
+    let not_an_elf = io::Error::new(io::ErrorKind::InvalidInput, "not a 64-bit ELF");
+    match maybe_elf {
+        goblin::Object::Elf(elf) => {
+            if elf.is_64 {
+                Ok(elf)
+            } else {
+                Err(not_an_elf)
+            }
+        }
+        _ => Err(not_an_elf),
+    }
+}
+
+pub fn load_elf64(image: &[u8], address_space: &mut AddressSpace) -> Result<VirtAddr, io::Error> {
+    let elf = parse_elf64(&image)?;
+    let headers: Vec<&ProgramHeader> = elf
+        .program_headers
+        .iter()
+        .filter(|hdr| hdr.p_type == PT_LOAD && hdr.p_memsz > 0)
+        .collect();
+
+    for ph in headers {
+        let guest_phys_start = PhysAddr::new(ph.p_paddr);
+        let host_virt_start = address_space.to_host_virt_addr(guest_phys_start)?;
+        let destination = unsafe {
+            std::slice::from_raw_parts_mut(host_virt_start.as_mut_ptr::<u8>(), ph.p_memsz as _)
+        };
+
+        destination[..ph.p_filesz as usize]
+            .copy_from_slice(&image[ph.p_offset as usize..(ph.p_offset + ph.p_filesz) as usize]);
+        unsafe {
+            if ph.p_memsz > ph.p_filesz {
+                std::ptr::write_bytes(
+                    &mut destination[ph.p_filesz as usize] as *mut u8,
+                    0u8,
+                    ph.p_memsz as usize - ph.p_filesz as usize,
+                );
+            }
+        }
+    }
+
+    let virt_entry = address_space.to_host_virt_addr(PhysAddr::new(elf.entry))?;
+    Ok(virt_entry)
+}
```

Comment:
Do you have interest in sharing [this code](https://github.com/enarx/enarx/blob/master/enarx-keep-sgx/src/component.rs)? It currently has one SGX data type, but we could probably fix that. I can think of lots of reasons why this might be beneficial:

1. Common validation
2. Common error types
3. Common user error messages

Reply:
I'm experimenting with this right now, thanks!